### PR TITLE
CASR deduplication for StacktraceObservers

### PR DESCRIPTION
**Pull request by anfedotoff**

### CODE REVIEW

File: libafl/src/observers/casr.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        Ok(())
+    }
+
+    /// read ASAN output from the log file and parse it.
+    pub fn parse_asan_output_from_asan_log_file(&mut self, pid: i32) -> Result<(), Error> {
+        let log_path = format!("{ASAN_LOG_PATH}.{pid}");
+        let mut asan_output = File::open(Path::new(&log_path))?;
+
+        let mut buf = String::new();
+        asan_output.read_to_string(&mut buf)?;
```
Comment (SweetVishnya) [inline libafl/src/observers/casr.rs:56]: utf8 lossy?

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

**Review by SweetVishnya — APPROVED** [review]

File: fuzzers/backtrace_baby_fuzzers/command_executor/src/main.rs [hunk 2]
```diff
@@ -24,13 +24,21 @@ use libafl::{
...
     mutators::scheduled::{havoc_mutations, StdScheduledMutator},
-    observers::{get_asan_runtime_flags, AsanBacktraceObserver, StdMapObserver},
+    observers::{get_asan_runtime_flags, StdMapObserver},
     schedulers::QueueScheduler,
     stages::mutational::StdMutationalStage,
     state::StdState,
     Error,
 };
 
+cfg_if::cfg_if! {
```
Comment (domenukk) [inline fuzzers/backtrace_baby_fuzzers/command_executor/src/main.rs:34]: You won't need a crate here, you can do
```rust
#[cfg(feature = "casr")]
use libafl::observers::CasrAsanBacktraceObserver;
#[cfg(not(feature = "casr"))]
use libafl::observers::AsanBacktraceObserver;
```
or maybe even just always enable AsanBacktraceObserver, even for the casr feature - why not?

**Comment by anfedotoff** [comment]
@domenukk, suddenly I realized this: CasrAsanObserver and AsanBacktraceObserver are the same expect for this function:
https://github.com/AFLplusplus/LibAFL/blob/a01863696e186c0d57acdc6789e16d59ce96c01f/libafl/src/observers/stacktrace.rs#L226
Maybe I use `feature = "casr"` / `not(feature = "casr")` to switch between this functions? This will reduce code significantly.
The same thing I could do for:
https://github.com/AFLplusplus/LibAFL/blob/a01863696e186c0d57acdc6789e16d59ce96c01f/libafl/src/observers/stacktrace.rs#L28
So all stacktrace observes could switch between hashing algorithms. All examples should work without any code modification! What do you think?

File: libafl/src/observers/stacktrace.rs [hunk 3]
```diff
@@ -233,6 +292,22 @@ impl AsanBacktraceObserver {
         self.update_hash(hash);
     }
 
+    #[cfg(feature = "casr")]
+    /// parse ASAN error output emited by the target command and compute the hash
+    pub fn parse_asan_output(&mut self, output: &str) {
+        let mut hash = 0;
+        if let Ok(st_vec) = AsanStacktrace::extract_stacktrace(output) {
+            if let Ok(mut stacktrace) = AsanStacktrace::parse_stacktrace(&st_vec) {
+                init_ignored_frames!("rust", "cpp", "go");
```
Comment (SweetVishnya) [inline libafl/src/observers/stacktrace.rs:301]: same here, we should init this only once

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: libafl/src/observers/stacktrace.rs [hunk 4]
```diff
@@ -45,6 +68,41 @@ pub fn collect_backtrace() -> u64 {
...
+#[must_use]
+pub fn collect_backtrace() -> u64 {
+    let mut b = Backtrace::new_unresolved();
+    if b.frames().is_empty() {
+        return 0;
+    }
+    b.resolve();
+    init_ignored_frames!("rust", "cpp");
+    let mut strace = Stacktrace::new();
+    for frame in &b.frames()[1..] {
```
[the remark this answers is not in the document]
Comment (anfedotoff) [reply libafl/src/observers/stacktrace.rs:82]: I suppose, no. According to this: https://docs.rs/backtrace/latest/backtrace/struct.Backtrace.html#method.frames

**Comment by anfedotoff** [comment]
I think now it's ready! @domenukk, please, could you have a look?

File: libafl/src/observers/stacktrace.rs [hunk 5]
```diff
@@ -66,13 +125,30 @@ pub struct BacktraceObserver<'a> {
...
+
+    #[cfg(feature = "casr")]
     /// Creates a new [`BacktraceObserver`] with the given name.
     #[must_use]
     pub fn new(
         observer_name: &str,
         backtrace_hash: &'a mut Option<u64>,
         harness_type: HarnessType,
     ) -> Self {
+        init_ignored_frames!("rust", "cpp");
```
Comment (SweetVishnya) [inline libafl/src/observers/stacktrace.rs:151]: why not adding go here?

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

**Review by SweetVishnya — APPROVED** [review]

**Comment by anfedotoff** [comment]
Btw, I tested AsanBacktraceObserver with CASR on some real world example https://github.com/tfussell/xlnt. For this purpose I modified `forkserver_simple` and cherry-picked this fix #1189. It seems to work! If we need this example I could prepare.

**Comment by tokatoka** [comment]
looks good thank you!

**Pull request merged**